src/vcpkg/base/downloads.cpp

```diff
@@ -1152,144 +531,102 @@ namespace vcpkg
-                    return DownloadPrognosis::OtherError;
-                }
+        std::error_code ec;
+        auto fileptr = std::make_unique<WriteFilePointer>(download_path_part_path, Append::NO, ec);
+        if (ec)
+        {
+            context.report_error(format_filesystem_call_error(ec, "fopen", {download_path_part_path}));
+            return DownloadPrognosis::OtherError;
+        }
 
-                auto authority = StringView{*maybe_authority}.substr(2);
-                // This check causes complex URLs (non-default port, embedded basic auth) to be passed down to
-                // curl.exe
-                if (Strings::find_first_of(authority, ":@") == authority.end())
+        curl_slist* request_headers = nullptr;
+        for (auto&& header : headers)
+            request_headers = curl_slist_append(request_headers, header.c_str());
+
+        auto curl = curl_easy_init();
+        if (!curl) Checks::unreachable(VCPKG_LINE_INFO);
+
+        set_common_curl_options(curl, url_encode_spaces(raw_url).c_str(), request_headers);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &write_file_callback);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, fileptr.get());
+        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 0L); // enable progress
+        curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, &progress_callback);
+        curl_easy_setopt(curl, CURLOPT_XFERINFODATA, &machine_readable_progress);
+
+        // Retry on transient errors:
+        // Transient error means either: a timeout, an FTP 4xx response code or an HTTP 408, 429, 500, 502, 503 or 504
+        // response code.
+        bool curl_success = false;
+        bool should_retry = true;
+        size_t retries_count = 0;
+
+        using namespace std::chrono_literals;
+        static constexpr std::array<std::chrono::seconds, 3> retry_delay = {0s, 1s, 2s};
+        do
+        {
+            // blocking transfer
+            should_retry = false;
+            std::this_thread::sleep_for(retry_delay[retries_count++]);
+            auto curl_code = curl_easy_perform(curl);
+            if (curl_code == CURLE_OK)
+            {
+                long response_code = -1;
+                if (CURLE_OK == curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code))
                 {
-                    if (!download_winhttp(context,
-                                          machine_readable_progress,
-                                          fs,
-                                          download_path_part_path,
-                                          *split_uri_view,
-                                          sanitized_url))
+                    if ((response_code >= 200 && response_code < 300) || response_code == 0)
                     {
-                        return DownloadPrognosis::NetworkErrorProxyMightHelp;
+                        curl_success = true;
+                        break;
```

Comment:
Note for other reviewers: that this `break;` is not meaningful as `should_retry` is false. *Probably* still worth keeping here but it doesn't *really* introduce funny exit paths.